File: ios/Plugin/KlaviyoSDKCapacitorPlugin.swift
import Foundation
import Capacitor
import KlaviyoSwift
import UserNotifications

@objc(KlaviyoSDKCapacitorPlugin)
public class KlaviyoSDKCapacitorPlugin: CAPPlugin, UNUserNotificationCenterDelegate {
  private let implementation = KlaviyoSDKCapacitor()
  
  public override func load() {
    // Listen for the notification from the AppDelegate
    UNUserNotificationCenter.current().delegate = self
  }
  
  @objc func initSDK(_ call: CAPPluginCall) {
    let value = call.getString("klaviyoKey") ?? ""
    KlaviyoSDK().initialize(with: value)
    call.resolve([
      "result": true
    ])
  }
  
  @objc func setUser(_ call: CAPPluginCall) {
    let email = call.getString("email") ?? ""
    let firstName = call.getString("firstName") ?? ""
    let lastName = call.getString("lastName") ?? ""

    let profile = Profile(email: email,  firstName: firstName,  lastName: lastName)

    KlaviyoSDK().set(profile: profile)

    call.resolve([
      "result": true
    ])
  }
  
  @objc func setPushToken(_ call: CAPPluginCall) {
    let token = call.getString("token") ?? ""
    KlaviyoSDK().set(pushToken: token)
    call.resolve([
      "result": true
    ])
  }
  
  @objc func resetProfile(_ call: CAPPluginCall) {
    KlaviyoSDK().resetProfile()
    call.resolve([
      "result": true
    ])
  }
}

extension KlaviyoSDKCapacitorPlugin {
  // below method will be called when the user interacts with the push notification
  public func userNotificationCenter(
    _ center: UNUserNotificationCenter,
    didReceive response: UNNotificationResponse,
    withCompletionHandler completionHandler: @escaping () -> Void) {
      if #available(iOS 16.0, *) {
        UNUserNotificationCenter.current().setBadgeCount(UIApplication.shared.applicationIconBadgeNumber - 1)
      } else {
        UIApplication.shared.applicationIconBadgeNumber -= 1
      }
      
      // If this notification is Klaviyo's notification we'll handle it
      // else pass it on to the next push notification service to which it may belong
      let handled = KlaviyoSDK().handle(notificationResponse: response, withCompletionHandler: completionHandler)
      if !handled {
        completionHandler()
      }
    }
  
  // below method is called when the app receives push notifications when the app is the foreground
  public func userNotificationCenter(
    _ center: UNUserNotificationCenter,
    willPresent notification: UNNotification,
    withCompletionHandler completionHandler: @escaping (UNNotificationPresentationOptions) -> Void) {
      if #available(iOS 14.0, *) {
        completionHandler([.list, .banner])
      } else {
        completionHandler([.alert])
      }
    }
}

File: android/src/main/java/com/capacitor/klaviyosdk/plugin/KlaviyoSDKCapacitorPlugin.java
package com.capacitor.klaviyosdk.plugin;

import android.util.Log;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.klaviyo.analytics.Klaviyo;
import com.klaviyo.analytics.model.ProfileKey;

@CapacitorPlugin(name = "KlaviyoSDKCapacitor")
public class KlaviyoSDKCapacitorPlugin extends Plugin {

    private KlaviyoSDKCapacitor implementation = new KlaviyoSDKCapacitor();

    @PluginMethod
    public void initSDK(PluginCall call) {
        try {
            String key = call.getString("klaviyoKey");
            Klaviyo.INSTANCE.initialize(key, getActivity().getApplicationContext());
            JSObject ret = new JSObject();
            ret.put("result", key);
            call.resolve(ret);
        } catch (Error e) {
            JSObject ret = new JSObject();
            ret.put("result", false);
            call.resolve(ret);
        }
    }

    @PluginMethod
    public void setUser(PluginCall call) {
        try {
            String email = call.getString("email");
            String firstName = call.getString("firstName");
            Klaviyo.INSTANCE.setEmail(email).setProfileAttribute(ProfileKey.FIRST_NAME.INSTANCE, firstName);
            JSObject ret = new JSObject();
            ret.put("result", email);
            call.resolve(ret);
        } catch (Error e) {
            JSObject ret = new JSObject();
            ret.put("result", false);
            call.resolve(ret);
        }
    }

    @PluginMethod
    public void setPushToken(PluginCall call) {
        try {
            String token = call.getString("token");
            Klaviyo.INSTANCE.setPushToken(token);
            JSObject ret = new JSObject();
            ret.put("token", token);
            call.resolve(ret);
        } catch (Error e) {
            JSObject ret = new JSObject();
            ret.put("token", false);
            call.resolve(ret);
        }
    }

    @PluginMethod
    public void resetProfile(PluginCall call) {
        try {
            Klaviyo.INSTANCE.resetProfile();
            JSObject ret = new JSObject();
            Log.d("KlaviyoReset", "reset success");
            ret.put("result", true);
            call.resolve(ret);
        } catch (Error e) {
            JSObject ret = new JSObject();
            Log.d("KlaviyoReset", "reset failed");
            ret.put("result", false);
            call.resolve(ret);
        }
    }
}
